packages/txservice/src/transaction.ts

```diff
@@ -0,0 +1,216 @@
+      },
+      "Attempting to send transaction.",
+    );
+    const { response: _response, success } = await this.provider.sendTransaction(await this.fullTransaction);
+    if (!success) {
+      const error = _response as Error;
+      if (
+        this.responses.length >= 1 &&
+        (error.message.includes("nonce has already been used") ||
+          // If we get a 'nonce is too low' message, a previous tx has been mined, and ethers thought
+          // we were making another tx attempt with the same nonce.
+          error.message.includes("Transaction nonce is too low.") ||
+          // Another ethers message that we could potentially be getting back.
+          error.message.includes("There is another transaction with same nonce in the queue."))
+      ) {
+        this.nonceExpired = true;
+        this.log.info(
+          {
+            method,
+            ...fullTransaction,
+            gasPrice: fullTransaction.gasPrice.toString(),
+            error: error.message,
+          },
+          "Nonce already used.",
+        );
+      } else {
+        this.log.error({ method, error }, "Failed to send tx");
+        throw _response;
+      }
+    }
+    const response = _response as providers.TransactionResponse;
+
+    // Save nonce if applicable.
+    if (this.nonce === undefined) {
+      this.nonce = response.nonce;
+    }
+    // TODO: Should we be checking for wild case where this.nonce !== response.nonce?
+
+    // Add this response to our local response history.
+    this.responses.push(response);
+
+    this.log.info(
+      {
+        method,
+        hash: response.hash,
+        gas: (response.gasPrice ?? "unknown").toString(),
+        nonce: response.nonce,
+      },
+      "Tx submitted",
+    );
+  }
+
+  public async confirm() {
+    const { confirmationTimeout, confirmationTimeoutExtensionMultiplier } = this.config;
+    // A flag for marking when we have received at least 1 confirmation. We'll extend the wait period
+    // if this is the case.
+    let receivedConfirmation = false;
+    let attempts = 0;
+
+    // An anon fn to get the tx receipts for all responses.
+    // We must check for confirmation in all previous transactions. Although it's most likely
+    // that it's the previous one, any of them could have been confirmed.
+    const waitForReceipt = async (): Promise<providers.TransactionReceipt | undefined> => {
+      attempts += 1;
+      // Save all reverted receipts for a check in case our Promise.race evaluates to be undefined.
+      const reverted: providers.TransactionReceipt[] = [];
+      // Make a pool of promises for resolving each receipt call (once it reaches target confirmations).
+      const receipt = await Promise.race<any>(
+        this.responses
+          .map((response) => {
+            return new Promise(async (resolve) => {
+              const r = await this.provider.confirmTransaction(response.hash, confirmationTimeout);
```

Comment:
We should probably consider making the confirmation timeout specific to various chains

Reply:
Yes, the config should fully be chain-dependent, confirmation timeout, blocks to wait, etc.

Reply:
I will just remove confirmationTimeout from args here, and have it be a property of the provider instead, since the provider is chain-dependent